File: Graph programs/N-Queens/N-Queens.cpp
#include <iostream>
#include <vector>
#include <cstdlib> 
#include <ctime>    


const int MAX_N = 51;

void generate_board(int n, std::vector<int>& board)
{
    for (int row = 0; row < n; ++row)
        board[row] = std::rand() % n;
}


bool check_is_possible(int n)
{
    if (n == 2 || n == 3)
    {
        std::cout << "Not possible with n = " << n << '\n';
        return false;
    }
    return true;
}


void print_pos(const std::vector<int>& board)
{
    
    for (int i = 0; i < board.size(); ++i)
    {
        std::cout << board[i] << " ";
    }
    std::cout << '\n';
}

void print_board(const std::vector<int>& board) 
{
    int n = board.size();
    if (n >= MAX_N)
    {
        std::cout << "Board too large to display (" << n << "x" << n << ")\n";
        return;
    }

    for (int row = 0; row < n; ++row) {
        for (int col = 0; col < n; ++col)
        {
            if (board[col] == row)
                std::cout << "* ";
            else
                std::cout << "_ ";
        }
        std::cout << "\n";
    }
}

bool check_diagonal(const std::vector<int>& board)
{
    int n = board.size();
    for (int i = 0; i < n - 1; ++i)
    {
        for (int j = i + 1; j < n; ++j)
        {
            int row_diff = std::abs(i - j);
            int col_diff = std::abs(board[i] - board[j]);
            if (row_diff == col_diff)
                return false;
        }
    }
    return true; 
}

bool check_col(const std::vector<int>& board)
{
    int n = board.size();
    for (int i = 0; i < n - 1; ++i)
    {
        for (int j = i + 1; j < n; ++j)
        {
            if (board[i] == board[j])
                return false;
        }
    }
    return true;
}

int conflicts_for(const std::vector<int>& board, int row, int col) {
    int n = board.size();
    int conflicts = 0;

    for (int r = 0; r < n; ++r)
    {
        if (r == row)
            continue;

        int c = board[r];

        if (c == col)
        { 
            conflicts++;
            continue;
        }

        if (std::abs(r - row) == std::abs(c - col))
            conflicts++;
    }
    return conflicts;
}


static inline void collect_conflicted_rows(const std::vector<int>& board, std::vector<int>& out_rows)
{
    out_rows.clear();
    int n = board.size();

    for (int row = 0; row < n; ++row)
    {
        if (conflicts_for(board, row, board[row]) > 0)
            out_rows.push_back(row);
    }
}


int pick_best_col_random_tiebreak(const std::vector<int>& costs) {
    int n = costs.size();
    int best = costs[0];
    for (int c = 1; c < n; ++c) 
        best = std::min(best, costs[c]);

    std::vector<int> bucket;
    for (int c = 0; c < n; ++c)
        if (costs[c] == best) bucket.push_back(c);

    int idx = std::rand() % (int)bucket.size();
    return bucket[idx];
}


bool min_conflicts(std::vector<int>& board, int max_steps = 50000) {
    int n = board.size();
    if (n == 0)
        return true;

    std::vector<int> conflicted_rows;
    conflicted_rows.reserve(n);

    for (int step = 0; step < max_steps; ++step) {

        collect_conflicted_rows(board, conflicted_rows);
        if (conflicted_rows.empty()) return true;


        int row = conflicted_rows[std::rand() % (int)conflicted_rows.size()];


        std::vector<int> costs(n);
        for (int col = 0; col < n; ++col)
            costs[col] = conflicts_for(board, row, col);

        int best_col = pick_best_col_random_tiebreak(costs);

  
        board[row] = best_col;
    }

    return false; 
}


void print_final(std::vector<int>& board)
{
    int n = board.size();
    if (!check_is_possible(n))
        return;
    generate_board(n, board);
    std::cout << "\nFirst Positions:\n";
    print_pos(board);
    std::cout << "\nFirst board:\n";
    print_board(board);
    min_conflicts(board);
    std::cout << "\n\n\nFinal Positions (After Min-Conflicts):\n";
    print_board(board);
    std::cout << "\nFinal board(After Min-Conflicts):\n";
    print_pos(board);
}

int main()
{
    int n = 0;
    std::cin >> n;
    std::vector<int> board(n);
    print_final(board);
}

File: Graph programs/N-Queens/n-queens.py
import random

MAX_N = 51


def generate_board(n):
    return [random.randint(0, n - 1) for _ in range(n)]


def check_is_possible(n):
    if n in (2, 3):
        #print(f"Not possible with n = {n}")
        return False
    return True


def print_pos(board):
    print(" ".join(map(str, board)))


def print_board(board):
    n = len(board)
    if n >= MAX_N:
        print(f"Board too large to display ({n}x{n})")
        return

    for row in range(n):
        line = []
        for col in range(n):
            line.append("*" if board[col] == row else "_")
        print(" ".join(line))


def conflicts_for(board, row, col):
    n = len(board)
    conflicts = 0
    for r in range(n):
        if r == row:
            continue

        c = board[r]

        if c == col:
            conflicts += 1
            continue

        if abs(r - row) == abs(c - col):
            conflicts += 1

    return conflicts


def collect_conflicted_rows(board):
    conflicted_rows = []  

    for row in range(len(board)):
        conflicts = conflicts_for(board, row, board[row])

        if conflicts > 0:
            conflicted_rows.append(row)

    return conflicted_rows


def pick_best_col_random_tiebreak(costs):

    best = min(costs)
    best_cols = []
    for i in range(len(costs)):
        if costs[i] == best:
            best_cols.append(i)
    return random.choice(best_cols)


def min_conflicts(board, max_steps=50000):
    n = len(board)
    if n == 0:
        return True

    for step in range(max_steps):
        conflicted_rows = collect_conflicted_rows(board)
        if not conflicted_rows:
            return True  

        row = random.choice(conflicted_rows)

        costs = [conflicts_for(board, row, col) for col in range(n)]

        best_col = pick_best_col_random_tiebreak(costs)

        board[row] = best_col

    return False  


def print_final(board):
    n = len(board)
    if not check_is_possible(n):
        return

    board[:] = generate_board(n)

    print("\nFirst Positions:")
    print_pos(board)

    print("\nFirst board:")
    print_board(board)

    solved = min_conflicts(board)

    if solved:
        print("\n\nFinal Positions (After Min-Conflicts):")
        print_pos(board)
        print("\nFinal board (After Min-Conflicts):")
        print_board(board)
    else:
        print("\nFailed to find solution after max steps.")

def print_final_for_test(board):
    n = len(board)

    if not check_is_possible(n):
        print(-1)
        return
    
    if n > 700:
        print(-1)
        return

    board[:] = generate_board(n)

    solved = min_conflicts(board)
    if solved:
        #print_board(board)
        print_pos(board)
    else:
        print(-1)

def main():
    n = int(input().strip())
    board = [0] * n
    print_final_for_test(board)


if __name__ == "__main__":
    main()
